include/dsn/idl/thrift_helper.h

```diff
@@ -284,12 +285,44 @@ namespace dsn {
         return serialization_forwarder<TName>::unmarshall(iproto, val);
     }
 
+#define GET_THRIFT_TYPE_MACRO(cpp_type, thrift_type) \
+    inline ::apache::thrift::protocol::TType get_thrift_type(const cpp_type&)\
+    {\
+        return ::apache::thrift::protocol::thrift_type;\
+    }\
+
+    GET_THRIFT_TYPE_MACRO(bool, T_BOOL)
+    GET_THRIFT_TYPE_MACRO(int8_t, T_BYTE)
+    GET_THRIFT_TYPE_MACRO(uint8_t, T_BYTE)
+    GET_THRIFT_TYPE_MACRO(int16_t, T_I16)
+    GET_THRIFT_TYPE_MACRO(uint16_t, T_I16)
+    GET_THRIFT_TYPE_MACRO(int32_t, T_I32)
+    GET_THRIFT_TYPE_MACRO(uint32_t, T_I32)
+    GET_THRIFT_TYPE_MACRO(int64_t, T_I64)
+    GET_THRIFT_TYPE_MACRO(uint64_t, T_U64)
+    GET_THRIFT_TYPE_MACRO(double, T_DOUBLE)
+    GET_THRIFT_TYPE_MACRO(std::string, T_STRING)
+
+    template<typename T>
+    inline ::apache::thrift::protocol::TType get_thrift_type(const std::vector<T>&)
+    {
+        return ::apache::thrift::protocol::T_LIST;
+    }
+
+    template<typename T>
+    inline ::apache::thrift::protocol::TType get_thrift_type(const T&)
+    {
+        return ::apache::thrift::protocol::T_STRUCT;
+    }
+
     template<typename T>
     void marshall(binary_writer& writer, const T& val)
     {
         boost::shared_ptr< ::dsn::binary_writer_transport> transport(new ::dsn::binary_writer_transport(writer));
```

Comment:
this is my bad. But can we remove new/delete here but use on-stack vars here instead to improve the performance?